### PR TITLE
Add miscellaneous changes for benchmarking

### PR DESCRIPTION
**Pull request by CatherineSue**
* Algorithm
   * Fix entropy in tf/algos/npo.py. According to the PPO paper, the term
     `policy_ent_coeff * pol_ent` should be added to the policy gradient
     loss instead of the reward. By fixing this, the policy stddev will
     decrease if policy_ent_coeff is very small.
   * Also addressed issue #442.

 * Sampler
   * baselines logs average return of the recent 100 trajectories. Add
     info to garage.tf.BatchSampler. Note this is a temp fix to use
     benchmark.

 * Benchmark
    * Update the current benchmark to work properly. This includes:
      * Add AutoStopEnv wrapper. baselines doesn't set done=True in its
        VecNormEnv. To make the sampler behavior consistent with garage,
        the AutoStopEnv wrapper will set done to True after
        `max_path_length` steps.
      * Replace `trial` typo in test_benchmark_*.py.
      * Move tf.Graph() to parent function as baselines can't run
        multiple times in the same tf graph.
    * Add tensorboard log support to baselines. 
    * Currently, the test_benchmark_ppo and test_benchmark_trpo works.
      test_benchmark_ddpg and test_benchmark_her may not work. For further
      info please refer to issue #568. 

  * Regressor
    * Rename `step_size` to `max_kl_step` in tf/regressors to consistent
      with the naming convention garage uses.
    * Add name to baselines so that NPO can use its name in
     `_fit_baseline`.

Note this PR made some band-aid fix in order to make the benchmark work.
They should be addressed when the sampler and new logger is merged.

### CODE REVIEW

**Comment by CatherineSue** [comment]
#PPO Benchmark Result - 1
## Hyperparameters
|Hyperparameter        |Value  |         
|:------------- |:-------------|
| Dicount      | 0.99 |
| GAE parameter | 0.95 |
| Minibatch size | 32 |
| Num. epochs | 10 |
| Horizon | 2048 |
| Policy hidden sizes | (64, 64) |
| Baseline(value function) hidden sizes | (64, 64) |
| Baseline optimizer | PenaltyLbfgsOptimizer |
| Adam optimizer learning rate | 1e-3 |
| Clip range | 0.2 | 

Note that baselines seems to be unstable when the learning rate is 1e-3. This may be related to their policy-vf shared network architecture. Their default lr is 3e-4. I found that `learning_rate=3e-4` causes a slow convergence in garage. Since `learning_rate=1e-3` is the default setting of tf.train.Adamoptimizer, I stick with 1e-3.

## InvertedDoublePendulum-v2
InvertedDoublePendulum-v2_benchmark

## InvertedPendulum-v2
InvertedPendulum-v2_benchmark

## HalfCheetah-v2
HalfCheetah-v2_benchmark

## Hopper-v2
Hopper-v2_benchmark

## Walker2d-v2
Walker2d-v2_benchmark

Reacher and Swimmer is still running. I'll update later.

**Comment by CatherineSue** [comment]
# PPO Benchmark Result - 2
## Hyperparameters
|Hyperparameter        |Value  |         
|:------------- |:-------------|
| Dicount      | 0.99 |
| GAE parameter | 0.95 |
| Minibatch size | 32 |
| Num. epochs | 10 |
| Horizon | 2048 |
| Policy hidden sizes | (64, 64) |
| Baseline(value function) hidden sizes | (64, 64) |
| Baseline optimizer | **FirstOrderOptimizer** |
| Adam optimizer learning rate | 1e-3 |
| Clip range | 0.2 | 

Note that baselines seems to be unstable when the learning rate is 1e-3. This may be related to their policy-vf shared network architecture. Their default lr is 3e-4. I found that `learning_rate=3e-4` causes a slow convergence in garage. Since `learning_rate=1e-3` is the default setting of tf.train.Adamoptimizer, I stick with 1e-3.

## InvertedDoublePendulum-v2
InvertedDoublePendulum-v2_benchmark

## InvertedPendulum-v2
InvertedPendulum-v2_benchmark

## HalfCheetah-v2
HalfCheetah-v2_benchmark

## Hopper-v2
Hopper-v2_benchmark

## Walker2d-v2
Walker2d-v2_benchmark

Reacher and Swimmer is still running. I'll update later.

**Comment by CatherineSue** [comment]
## Entropy Test
Screen Shot 2019-03-08 at 1 06 21 PM

**Comment by ryanjulian** [comment]
These look great -- regarding entropy: there are several ways to implement this. The one in the PPO paper is just one. Sometimes it is an augmented loss and sometimes it is augmented reward (maximum entropy vs entropy-regularized). 

I would like to see NPO support both implementations, but it's better right now to have at least one which is known to work.

File: garage/tf/samplers/batch_sampler.py [hunk 1]
```diff
@@ -143,6 +147,7 @@ def process_samples(self, itr, paths):
         logger.record_tabular('AverageDiscountedReturn',
                               average_discounted_return)
         logger.record_tabular('AverageReturn', np.mean(undiscounted_returns))
+        logger.record_tabular('eprewmean', np.mean(self.eprewmean))
```
Comment (ryanjulian) [inline garage/tf/samplers/batch_sampler.py]: if we are going to log this, we should use our own naming convention (e.g. `EpisodeRewardMean`). I think there are already mappings between our names and baseline names in the benchmark scripts.

File: tests/benchmarks/test_benchmark_ddpg.py [hunk 2]
```diff
@@ -2,8 +2,8 @@
 This script creates a regression test over garage-DDPG and baselines-DDPG.
```
Comment (ryanjulian) [inline tests/benchmarks/test_benchmark_ddpg.py]: just a thought: perhaps we should add a test case to each benchmark which runs each benchmark for just 1 iteration, to make sure that all the benchmarks stay running using the CI.

we would still exclude the main benchmark from CI

File: tests/helpers.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import pickle
```
Comment (ryanjulian) [inline tests/helpers.py]: tests/wrappers.py?

there will probably be more of these.

File: garage/tf/samplers/batch_sampler.py (hunk 1 above)
Comment (CatherineSue) [reply garage/tf/samplers/batch_sampler.py]: Yes there are mappings for names. I used this naming to view the regression in tensorboard. I agree we should follow our naming convention. It would be better if we can move this extra info to somewhere else in the future. Because it only serves for benchmark.

File: tests/benchmarks/test_benchmark_ddpg.py (hunk 2 above)
Comment (CatherineSue) [reply tests/benchmarks/test_benchmark_ddpg.py]: Good idea. I believe Kevin has done a similar thing in his branch.

File: tests/helpers.py (hunk 3 above)
Comment (CatherineSue) [reply tests/helpers.py]: Good idea. I'll move it.

File: tests/benchmarks/test_benchmark_ddpg.py (hunk 2 above)
Comment (CatherineSue) [reply tests/benchmarks/test_benchmark_ddpg.py]: @cheng-kevin I assume the files under [this folder](https://github.com/rlworkgroup/garage/tree/Automated_Benchmarking_normalized/tests/benchmarks/continuous_integration) are for ci? Could you port them to this branch?

**Comment by CatherineSue** [comment]
I've updated the PPO benchmark results and updated notes. The DDPG benchmark is running.
I also addressed all the comments.

File: tests/benchmarks/test_benchmark_her.py [hunk 4]
```diff
@@ -64,18 +64,18 @@ def test_benchmark_her(self):
...
+                baselines_dir = trial_dir + "/baselines"
 
-                garage_csv = run_garage(env, seed, garage_dir)
+                with tf.Graph().as_default():
+                    garage_csv = run_garage(env, seed, garage_dir)
 
-                CACHED_ENVS.clear()
-                baselines_csv = run_baselines(env_id, seed, baselines_dir)
+                    CACHED_ENVS.clear()
+                    baselines_csv = run_baselines(env_id, seed, baselines_dir)
```
Comment (ahtsan) [inline tests/benchmarks/test_benchmark_her.py]: Do we need the AutoStop wrapper here?

File: tests/wrappers.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import gym
+
+
+class AutoStopEnv(gym.Wrapper):
+    """A env wrapper that stops rollout at step max_path_length."""
+
+    def __init__(self, env=None, env_name="", max_path_length=100):
+        if env_name:
+            super().__init__(gym.make(env_name))
```
Comment (ahtsan) [inline tests/wrappers.py]: Why do we need the env_name? Why can't we just pass the environment?

File: garage/tf/samplers/batch_sampler.py (hunk 1 above)
Comment (ryanjulian) [reply garage/tf/samplers/batch_sampler.py]: Perhaps just put them under "Extras/" or "Info/"?

File: tests/benchmarks/test_benchmark_her.py (hunk 4 above)
Comment (CatherineSue) [reply tests/benchmarks/test_benchmark_her.py]: No.

File: tests/wrappers.py (hunk 5 above)
Comment (CatherineSue) [reply tests/wrappers.py]: You can do both.

Comment (CatherineSue) [reply tests/wrappers.py]: I was referring to [this](https://github.com/rlworkgroup/garage/blob/master/garage/envs/base.py#L44).

Comment (ryanjulian) [reply tests/wrappers.py]: I think Anson was asking why you didn't use the standard `gym.Wrapper` interface [here](https://github.com/openai/gym/blob/master/gym/core.py#L209)

I also think that `gym.Wrapper` is preferable, but it's test code so I won't insist on any changes.

Comment (CatherineSue) [reply tests/wrappers.py]: Thanks for pointing it out. Yes I agree that one is better. I didn't pay much attention to this class because it was implemented before and I just pasted it here. My alter goal is that we can remove this when the new sampler is finished. Then this wrapper won't be required.

Comment (ryanjulian) [reply tests/wrappers.py]: then it doesn't seem we need to talk much about it, anyway :)
